arwen/host/asyncCall.go

```diff
@@ -264,3 +265,388 @@ func (host *vmHost) computeDataLengthFromArguments(function string, arguments []
+ *  done in two steps in order to correctly use all remaining gas. We first split the gas as specified by the developer,
+ *  then we save the storage, then we split again the gas to calls that leave this shard.
+ *
+ * returns a list of pending calls (the ones that should be processed on other hosts)
+ */
+func (host *vmHost) processAsyncInfo(asyncInfo *arwen.AsyncContextInfo) (*arwen.AsyncContextInfo, error) {
+	if len(asyncInfo.AsyncContextMap) == 0 {
+		return asyncInfo, nil
+	}
+
+	err := host.setupAsyncCallsGas(asyncInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, asyncContext := range asyncInfo.AsyncContextMap {
+		for _, asyncCall := range asyncContext.AsyncCalls {
+			if !host.canExecuteSynchronouslyOnDest(asyncCall.Destination) {
+				continue
+			}
+
+			procErr := host.processAsyncCall(asyncCall)
+			if procErr != nil {
+				return nil, err
+			}
+		}
+	}
+
+	pendingMapInfo := host.getPendingAsyncCalls(asyncInfo)
+	if len(pendingMapInfo.AsyncContextMap) == 0 {
+		return pendingMapInfo, nil
+	}
+
+	err = host.savePendingAsyncCalls(pendingMapInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	err = host.setupAsyncCallsGas(pendingMapInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, asyncContext := range pendingMapInfo.AsyncContextMap {
+		for _, asyncCall := range asyncContext.AsyncCalls {
+			if !host.canExecuteSynchronouslyOnDest(asyncCall.Destination) {
+				sendErr := host.sendAsyncCallToDestination(asyncCall)
+				if sendErr != nil {
+					return nil, sendErr
+				}
+			}
+		}
+	}
+
+	return pendingMapInfo, nil
+}
+
+/**
+ * processAsyncCall executes an async call and processes the callback if no extra calls are pending
+ */
+func (host *vmHost) processAsyncCall(asyncCall *arwen.AsyncGeneratedCall) error {
+	input, _ := host.createDestinationContractCallInput(asyncCall)
+	output, asyncMap, executionError := host.ExecuteOnDestContext(input)
+
+	pendingMap := host.getPendingAsyncCalls(asyncMap)
+	if len(pendingMap.AsyncContextMap) == 0 {
+		return host.callbackAsync(asyncCall, output, executionError)
+	}
+
+	return executionError
+}
+
+/**
+ * callbackAsync will execute a callback from an async call that was ran on this host and set it's status to resolved or rejected
+ */
+func (host *vmHost) callbackAsync(asyncCall *arwen.AsyncGeneratedCall, vmOutput *vmcommon.VMOutput, executionError error) error {
+	asyncCall.Status = arwen.AsyncCallResolved
+	callbackFunction := asyncCall.SuccessCallback
+	if vmOutput.ReturnCode != vmcommon.Ok {
+		asyncCall.Status = arwen.AsyncCallRejected
+		callbackFunction = asyncCall.ErrorCallback
+	}
+
+	callbackCallInput, err := host.createCallbackContractCallInput(
+		vmOutput,
+		asyncCall.Destination,
+		callbackFunction,
+		executionError,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Callback omits for now any async call - TODO: take into consideration async calls generated from callbacks
```

Comment:
The storage in the name referred to the fact that this function is sending a callback that it could not initiate on the initial contract execution (because it also generated a cross shard async call) an saved the callback information in storage so it can generate it when all it's own callbacks are executed. But I agree, it's a poor choice of name, especially now that after your feedback we no longer have `AsyncInitiator` and the signature bacame `sendStorageCallbackToDestination(callerAddress, returnData []byte)`.